test/dynamo/test_aot_autograd_cache.py

```diff
@@ -1,19 +1,260 @@
+
+        # A first call should miss in the cache.
+        self.assertEqual(fn(a, b), compiled_fn(a, b))
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_bypass"], 1)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_hit"], 0)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_saved"], 0)
+
+        # Clear FX graph cache: second call should also be a miss
+        self._clear_dynamo_and_codecache()
+
+        self.assertEqual(fn(a, b), compiled_fn(a, b))
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_bypass"], 2)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_hit"], 0)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_saved"], 0)
+
+    @inductor_config.patch("fx_graph_cache", True)
+    @functorch_config.patch({"enable_autograd_cache": True})
+    def test_autograd_function(self):
+        """
+        Tests autograd cache hits
+        """
+
+        def fn(a, b):
+            return a.sin() + b
+
+        a = torch.randn(25, requires_grad=True)
+        b = torch.randn(25, requires_grad=True)
+        a2 = a.detach().clone().requires_grad_(True)
+        b2 = b.detach().clone().requires_grad_(True)
+
+        compiled_fn = torch.compile(fn, backend="inductor")
+
+        # A first call should miss in the cache.
+        self.assertEqual(fn(a, b), compiled_fn(a2, b2))
+        fn(a, b).sum().backward()
+        compiled_fn(a2, b2).sum().backward()
+        self.assertEqual(a.grad, a2.grad)
+        self.assertEqual(b.grad, b2.grad)
+
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_miss"], 1)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_hit"], 0)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_saved"], 1)
+
+        # Reset all tensors
+        a = torch.randn(25, requires_grad=True)
+        b = torch.randn(25, requires_grad=True)
+        a2 = a.detach().clone().requires_grad_(True)
+        b2 = b.detach().clone().requires_grad_(True)
+
+        # A second call should hit. (First reset so in-memory guards
+        # don't prevent compilation).
+        self._clear_dynamo_and_codecache()
+        self.assertEqual(fn(a, b), compiled_fn(a2, b2))
+        fn(a, b).sum().backward()
+        compiled_fn(a2, b2).sum().backward()
+        self.assertEqual(a.grad, a2.grad)
+        self.assertEqual(b.grad, b2.grad)
+
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_miss"], 1)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_hit"], 1)
+        self.assertEqual(counters["aot_autograd"]["autograd_cache_saved"], 1)
+
+
+    @largeTensorTest("64GB", device=GPU_TYPE)
+    @parametrize("device", (GPU_TYPE,))
+    @parametrize("dtype", (torch.float16, torch.bfloat16))
+    @inductor_config.patch("fx_graph_cache", True)
+    @functorch_config.patch({"enable_autograd_cache": True})
+    def test_autograd_inductor_guards(self, device, dtype):
```

Comment:
TODO: if you do this test but also do autograd, we fail to serialize/deserialize the traced_tangents field of ViewAndMutationMeta because the FakeTensor we're serializing has symints in it: 
```
  File "/data/users/jjwu/a/pytorch/torch/_functorch/_aot_autograd/autograd_cache.py", line 371, in _fake_tensor_from_meta
    empty = torch.empty_strided(
  File "/data/users/jjwu/a/pytorch/torch/fx/experimental/sym_node.py", line 162, in maybe_as_int
    if self.expr.is_number:
  File "/data/users/jjwu/a/pytorch/torch/fx/experimental/sym_node.py", line 131, in expr
    return self.shape_env.replace(self._expr)
  File "/data/users/jjwu/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 1428, in wrapper
    if prior_version != self._version_counter:
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AttributeError: 'ShapeEnv' object has no attribute '_version_counter'
```
I'll probably change the serialization of ViewAndMutationMeta a bit to work around this

Reply:
Fixed in followup PR https://github.com/pytorch/pytorch/pull/127618